calyx/src/analysis/reaching_defns.rs

```diff
@@ -0,0 +1,422 @@
+            let (defs, par_killed): (Vec<DefSet>, Vec<KilledSet>) = stmts
+                .iter_mut()
+                .map(|ctrl| {
+                    build_reaching_def(
+                        ctrl,
+                        reach.clone(),
+                        KilledSet::new(),
+                        rd,
+                        counter,
+                    )
+                })
+                .unzip();
+
+            let global_killed = par_killed
+                .iter()
+                .fold(KilledSet::new(), |acc, set| &acc | set);
+
+            let par_exit_defs = defs
+                .iter()
+                .zip(par_killed.iter())
+                .map(|(defs, kills)| {
+                    defs.kill_from_hashset(&(&global_killed - kills))
+                })
+                .fold(DefSet::new(), |acc, element| &acc | &element);
+            (par_exit_defs, &global_killed | &killed)
+        }
+        ir::Control::If(ir::If {
+            tbranch,
+            fbranch,
+            cond,
+            ..
+        }) => {
+            let mut fake_enable = ir::Control::Enable(ir::Enable {
+                attributes: ir::Attributes::default(),
+                group: Rc::clone(cond),
+            });
+            let (post_cond_def, post_cond_killed) = build_reaching_def(
+                &mut fake_enable,
+                reach,
+                killed,
+                rd,
+                counter,
+            );
+            let (t_case_def, t_case_killed) = build_reaching_def(
+                tbranch,
+                post_cond_def.clone(),
+                post_cond_killed.clone(),
+                rd,
+                counter,
+            );
+            let (f_case_def, f_case_killed) = build_reaching_def(
+                fbranch,
+                post_cond_def,
+                post_cond_killed,
+                rd,
+                counter,
+            );
+            (&t_case_def | &f_case_def, &t_case_killed | &f_case_killed)
+        }
+        ir::Control::While(ir::While { cond, body, .. }) => {
+            let mut fake_enable = ir::Control::Enable(ir::Enable {
+                attributes: ir::Attributes::default(),
+                group: Rc::clone(cond),
+            });
+            let (post_cond_def, post_cond_killed) = build_reaching_def(
+                &mut fake_enable,
+                reach,
+                killed,
+                rd,
+                counter,
+            );
+
+            let (round_1_def, round_1_killed) = build_reaching_def(
+                body,
+                post_cond_def,
+                post_cond_killed,
+                rd,
+                counter,
+            );
+            let (post_cond2_def, post_cond2_killed) = build_reaching_def(
+                &mut fake_enable,
+                round_1_def,
+                round_1_killed,
+                rd,
+                counter,
+            );
+            // Twice as nice?
```

Comment:
Can you add a comment explaining why this is run twice? I assume this is to calculate the fixedpoint over the loop body.